### PR TITLE
cli: add buildtree checkout command

### PR DESCRIPTION
**Pull request by ownik**
This PR adds a "--use-buildtrees" option to "bst artifacts checkout", similar to the existing option in "bst shell".

The goal is to make it possible to check out build artifacts from the sandbox build tree when it is available.

Would be okay with adding this functionality, or are there reasons not to expose it through "bst artifacts checkout"? Or is there another way to get build artifacts from sandbox now?

### CODE REVIEW

**Comment by ownik** [comment]
@BenjaminSchubert @abderrahim @cs-shadow @gtristan @juergbi I've made changes to verify that buildtree exists in the cache, similar to the shell command. Could you please review, please?

**Comment by abderrahim** [comment]
Looks like this is fixing https://github.com/apache/buildstream/issues/1822, please link the issue in your PR. Also please add a test for this behaviour.

I'll try to review it today.

**Comment by abderrahim** [comment]
A few things need to be considered in the design of this feature:
* Someone recently suggested to implement this as a flag to `source checkout` rather than `artifact checkout`. It kinda makes sense that the build tree is related to the source, but it's also stored as part of the artifact. Maybe it makes sense to move it to a separate toplevel command? something like `bst buildtree checkout`?
* This is currently checking out the buildroot rather than what is traditionally considered a "build tree". This means it can work with both build elements and non-build-elements, but it also means we get a whole sandbox with the dependencies, the build tree, and even the artifact (if the build succeeded) which is probably not what people asking for the build tree are looking for.
* if we go with a new toplevel command, we might want to expose options to checkout either the build tree or the build root.

**Comment by ownik** [comment]
@abderrahim Thanks for the feedback! I've reworked the implementation into a new `bst buildtree checkout` command. By default it checks out the buildtree, and a `--buildroot` option can be used to export the entire buildroot instead.

I've also added an integration test covering the new command and link the issue in PR.

**Comment by ownik** [comment]
@abderrahim Any chance to review this PR?

File: src/buildstream/_stream.py [hunk 1]
```diff
@@ -916,6 +901,74 @@ def artifact_delete(self, targets, *, selection=_PipelineSelection.NONE):
...
+            connect_artifact_cache=True,
+            artifact_remotes=artifact_remotes,
+            ignore_project_artifact_remotes=ignore_project_artifact_remotes,
+        )
+
+        # self.targets contains a list of the loaded target objects
+        # if we specify --deps build, Stream._load() will return a list
+        # of build dependency objects, however, we need to prepare a sandbox
+        # with the target (which has had its appropriate dependencies loaded)
+        element: Element = self.targets[0]
```
Comment (abderrahim) [inline src/buildstream/_stream.py]: Does this make sense for the build tree? Either we're checking out just the buildtree, and then dependencies don't make sense, or we check out the whole buildroot and it already contains the build dependencies.

Am I missing something?

File: src/buildstream/_stream.py [hunk 2]
```diff
@@ -916,6 +901,74 @@ def artifact_delete(self, targets, *, selection=_PipelineSelection.NONE):
...
+
+        self._check_location_writable(location, force=force, tar=tar)
+
+        # Check whether the required elements are cached, and then
+        # try to pull them if they are not already cached.
+        #
+        self.query_cache(elements)
+        self._pull_missing_artifacts(elements)
```
Comment (abderrahim) [inline src/buildstream/_stream.py]: I think we only need to pull the artifact for the element in question. No need for dependencies here.

as an aside, I wonder if we should try to override the `pull-buildtrees` configuration in this case: if we're pulling to extract a build tree, we should probably try to pull the buildtree even if we usually don't.

File: src/buildstream/_stream.py [hunk 3]
```diff
@@ -916,6 +901,74 @@ def artifact_delete(self, targets, *, selection=_PipelineSelection.NONE):
...
+
+        self._check_buildtree(element)
+
+        try:
+            artifact = element._get_artifact()
+            virdir = artifact.get_buildroot() if buildroot else artifact.get_buildtree()
```
Comment (abderrahim) [inline src/buildstream/_stream.py:990]: Another thing we need to be aware of (and also give a suggestion about to the user) is that not all elements have a build tree: only build elements do.

So it's not enough to check for the build root, we also need to check for the build tree. Then depending on the availability and the user request, we can suggest an alternative to the user.

| buildtree available | buildroot available | user requested | result |
| --- | --- | --- | --- |
| Y | Y | build tree | extract build tree |
| Y | Y | build root | extract build root |
| N | Y | build tree | error + suggest extract build root |
| N | Y | build root | extract build root |
| N | N | build tree | error |
| N | N | build root | error |

File: tests/integration/buildtree.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+#
+# Verify fail cases when checkout buildtree
+#
+@pytest.mark.datafiles(DATA_DIR)
+@pytest.mark.skipif(not HAVE_SANDBOX, reason="Only available with a functioning sandbox")
+def test_buildtree_checkout_fail(cli, datafiles):
+    project = str(datafiles)
+    element_name = "build-shell/buildtree.bst"
+    checkout = os.path.join(cli.directory, "checkout")
+    tar = os.path.join(cli.directory, "source-checkout.tar")
```
Comment (abderrahim) [inline tests/integration/buildtree.py]:
```suggestion
    tar = os.path.join(cli.directory, "buildtree.tar")
```

File: tests/integration/buildtree.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+DATA_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "project")
+
+#
+# Verify fail cases when checkout buildtree
+#
+@pytest.mark.datafiles(DATA_DIR)
+@pytest.mark.skipif(not HAVE_SANDBOX, reason="Only available with a functioning sandbox")
+def test_buildtree_checkout_fail(cli, datafiles):
+    project = str(datafiles)
+    element_name = "build-shell/buildtree.bst"
```
Comment (abderrahim) [inline tests/integration/buildtree.py:41]: We should also test the buildtree of a script element (using `build-shell/script.bst`). This is the corner case I mentioned above where we have a build root but not a build tree.

**Review by abderrahim — COMMENTED** [review]
Thanks for the contributions. See my comments below.

File: src/buildstream/_stream.py (hunk 1 above)
Comment (ownik) [reply src/buildstream/_stream.py]: Sorry for the delayed response, and thanks for pointing this out.

Yes, you're right. I've changed this to `element: Element = elements[0]`, is it ok?

File: src/buildstream/_stream.py (hunk 2 above)
Comment (ownik) [reply src/buildstream/_stream.py]: Done 👍

File: src/buildstream/_stream.py (hunk 3 above)
Comment (ownik) [reply src/buildstream/_stream.py:990]: Thanks for pointing this out. I've accounted for all of the cases you described and added tests covering them.

Could you please take another look when you have a chance?

File: tests/integration/buildtree.py (hunk 4 above)
Comment (ownik) [reply tests/integration/buildtree.py]: Fixed

File: tests/integration/buildtree.py (hunk 5 above)
Comment (ownik) [reply tests/integration/buildtree.py:41]: i've added a test case which uses `build-shell/script.bst`
